Cms: new flag and option to omit signing time attribute

pjuhasz opened the pull request:
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [ ] tests are added or updated

These commits implement feature request #15777.

It was more complicated than I've first thought, I had to add a new member to the SignerInfo struct, but I hope that's acceptable.

I've updated the documentation and the cms command line application as well (in a separate commit). I haven't found any relevant tests.

slontis reviewed (commented):
This will need a test to be accepted.

Note this is changing the flow in
cms_SignerInfo_content_sign() which would have run through this branch
(And this could cause some potential issues)

if (CMS_signed_get_attr_count(si) >= 0)

t8m commented:
Too late for 3.0

pjuhasz commented:
> Note this is changing the flow in `cms_SignerInfo_content_sign()`

It isn't, if I'm reading it right. The signing time attribute is added (or, when my modification is in effect, not added) in `CMS_SignerInfo_sign()`. And `CMS_SignerInfo_sign()` is called _from_ `cms_SignerInfo_content_sign()`, inside that ```if (CMS_signed_get_attr_count(si) >= 0) {``` block. So that line comes earlier in the execution and my flag can have no effect on it.

openssl-machine commented:
This PR has been closed.  It was waiting for a CLA for 270 days.

t8m commented:
This cannot be merged without signed CLA from the author (@pjuhasz)

pjuhasz commented:
I'd be willing to provide one, but I don't know how.

FdaSilvaYY commented:
> I'd be willing to provide one, but I don't know how.

https://openssl-library.org/policies/cla/index.html

pjuhasz commented:
- The automatic check for "check style for each commit" has failed, but all it says is "Error:". Now what?
- There is a "hold: tests needed" label on the PR. What specific tests are needed, where?

t8m commented:
> * The automatic check for "check style for each commit" has failed, but all it says is "Error:". Now what?

You sometimes need to click on the failed CI step again to see the details. It's there.
 
>     * There is a "hold: tests needed" label on the PR. What specific tests are needed, where?

At least some tests are needed - it would be sufficient to test the app that the newly added option works as expected.

pjuhasz commented:
I've added a test, and fixed the coding style issues the tool was complaining about (with the exception of one that might be a false positive).
I've also rebased the branch to current master, I hope that's not a problem.
I'm also working on the CLA - finding a scanner is not trivial in this day and age.

t8m commented:
> I'm also working on the CLA - finding a scanner is not trivial in this day and age.

You can take a photo of the document with a mobile phone.

t8m commented on crypto/cms/cms_sd.c:
```diff
@@ -456,6 +457,14 @@ CMS_SignerInfo *CMS_add1_signer(CMS_ContentInfo *cms,
                 goto err;
             }
         }
+        if (flags & CMS_NO_SIGNING_TIME) {
```
Please use: `(flags & CMS_NO_SIGNING_TIME) != 0`

t8m commented on crypto/cms/cms_sd.c:
```diff
@@ -839,7 +848,8 @@ int CMS_SignerInfo_sign(CMS_SignerInfo *si)
         return 0;
 
-    if (CMS_signed_get_attr_by_NID(si, NID_pkcs9_signingTime, -1) < 0) {
+    if ((!si->omit_signing_time)
+        && (CMS_signed_get_attr_by_NID(si, NID_pkcs9_signingTime, -1) < 0)) {
```
Please drop the extraneous parentheses

t8m commented on crypto/cms/cms_sd.c:
```diff
@@ -848,7 +848,7 @@ int CMS_SignerInfo_sign(CMS_SignerInfo *si)
 
-    if ((!si->omit_signing_time)
+    if (!si->omit_signing_time
         && (CMS_signed_get_attr_by_NID(si, NID_pkcs9_signingTime, -1) < 0)) {
```
These parentheses are also superfluous

pjuhasz commented:
CLA sent.

t8m commented:
One last thing. Could you please add a CHANGES.md entry?

pjuhasz commented:
Done

beldmit reviewed (approved):
LGTM

bukka commented on apps/cms.c:
```diff
@@ -186,6 +187,8 @@ const OPTIONS cms_options[] = {
      "Don't include signer's certificate when signing"},
     {"noattr", OPT_NOATTR, '-', "Don't include any signed attributes"},
     {"nosmimecap", OPT_NOSMIMECAP, '-', "Omit the SMIMECapabilities attribute"},
+    {"no_signing_time", OPT_NO_SIGNING_TIME, '-',
```
Seems slightly inconsistent naming with other `no` flags (the undersores :) ) - `noattr`,  `nosmimecap`... So maybe `nosigningtime` would be more consistent. But they are some other with underscores so one could choose both. I don't really mind - just mentioning that if it was considered...

pjuhasz replied:
Well, naming is one of the hard problems of computer science :)
I did consider both versions. I eventually decided on the underscores version because it is consistent with the API flag, and as you say, there are already other options with underscores in them.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Squashed and merged to the master branch. Thank you for your contribution.
